[TP] Add wildcard support

kwen2501 opened the pull request:
Adding wildcard support for TP's `parallelize_module` API.

Example patterns:
`layers.*.linear`: any characters
`layers.?.linear`: single character
`layers.[1-2]`: digit range, matches `layers.1` and `layers.2`

Example use case:
A model have multiple layers, and we want to parallelize the linear module `lin` inside each layer.
```
model_tp = parallelize_module(
    model,
    device_mesh,
    {
        "layers.*.lin": ColwiseParallel(),
    },
)
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123218
* #123199
* __->__ #122968
* #122919



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

kwen2501 committed:
[TP] Add wildcard support

[ghstack-poisoned]

kurman commented:
I wonder if jq or xpath style spec would improve UX given the hierarchical nature? It comes with a higher complexity though.

XilunWu commented:
Users need be aware that the patterns in the dict param must be mutually exclusive otherwise the repetitive `parallelize_module` may cause issue.

XilunWu reviewed (approved):
LGTM

wz337 reviewed (approved):
LGTM! Could we add some tests to demonstrate the usage as well?

kwen2501 commented:
@wz337 Thanks for the review! The tests are in a stacked PR: #123101

wanchaol reviewed (changes requested):
Can you fold the tests PR to this PR? I think every new feature PR should come with tests in the PR itself, not a separate PR

kwen2501 committed:
Update on "[TP] Add wildcard support"

Adding wildcard support for TP's `parallelize_module` API.

Example patterns:
`layers.*.linear`: any characters
`layers.?.linear`: single character
`layers.[1-2]`: digit range, matches `layers.1` and `layers.2`

Example use case:
A model have multiple layers, and we want to parallelize the linear module `lin` inside each layer.
```
model_tp = parallelize_module(
    model,
    device_mesh,
    {
        "layers.*.lin": ColwiseParallel(),
    },
)
```




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

kwen2501 commented:
@wanchaol Done!

wanchaol commented on torch/testing/_internal/distributed/_tensor/common_dtensor.py:
```diff
@@ -78,6 +78,17 @@ def reset_parameters(self):
 
 
+class MLPStacked(nn.Module):
+    def __init__(self, device):
```
nit: I think it would be nice to have a `num_layers` arg (can have a default) to control how many MLP layers this stacked MLP construct

wanchaol commented on test/distributed/tensor/parallel/test_parallelize_api.py:
```diff
@@ -226,6 +228,70 @@ def test_prepare_module_output(self):
+            model_tp,
+            device_mesh,
+            {
+                "layers.*.net?": ColwiseParallel(output_layouts=Replicate()),
```
wondering can we do a e2e col + row test here?
```
{
    "layers.*.net[1]": ColwiseParallel(),
    "layers.*.net[2]": RowwiseParallel()
}
```

wanchaol reviewed (approved):
looks great, thanks for addressing comments!

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge
